---
title: "Comparison of CTCF consensus found in the (mutation-induced) Gained, Lost and stable binding sites: a descriptive analysis"
author: "by Kaiqiong Zhao"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: false
    number_sections: true
    toc_depth: 4
    keep_md: true
editor_options: 
  chunk_output_type: console
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, fig.width = 10, fig.height = 4) 
```



```{r}
library(magrittr)
library(rGADEM)
library(ggplot2)
library(reshape2)
library(MotIV)
library(BSgenome.Hsapiens.UCSC.hg19)
```

# Overview

The previous report, "(2)Motif_discovery_rGADEM.html", has shown that the most significant motif indentified in each of the three clusters, **Gained**, **Lost** and **Stable**, is the canonical CTCF motif. In this report, I will look into these identified CTCF-like censensus in each cluster and compare their binding site orientation, and their nucleotide distribution in windows with different lengths centered at the mid-point of the canonical CTCF motif, i.e the 10th nucleotide ('G') in the CTCF logo.

```{r, fig.width=10, fig.h=8}
path<- "/scratch/greenwood/kaiqiong.zhao/Old_home/kaiqiong.zhao/R/x86_64-redhat-linux-gnu-library/3.5/rGADEM/extdata/jaspar2009.txt"

seededPwm<-readPWMfile(path)
#grep("STAT1",names(seededPwm))

spwm = seededPwm$CTCF
seqLogo(apply(spwm, 2, function(x){x/sum(x)}))
```

Fig 1: sequence logo of the canonical CTCF motif


```{r}
load("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/WitcherKI/2-GADEM_top_1000/GADEM_up_1000.RData")
gadem_up <- gadem
load("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/WitcherKI/2-GADEM_top_1000/GADEM_down_1000.RData")
gadem_down <- gadem
load("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/WitcherKI/2-GADEM_top_1000/GADEM_null_1000.RData")
gadem_null <- gadem

motif11 <- gadem_up@motifList[[1]]
motif21 <- gadem_down@motifList[[1]]
motif31 <- gadem_null@motifList[[1]]
```

# CTCF binding site orientations in the three cluster

```{r}

summary_motif <- function(motif1){
seqID_vec <- rep(NA, length(motif1@alignList))
seq_vec <- strand_vec <- pval <- seqID_vec
 
for ( i in 1:length(seqID_vec)){
  seqID_vec[i] <- motif1@alignList[[i]]@seqID
  pval[i] <- motif1@alignList[[i]]@pval
  strand_vec[i] <- motif1@alignList[[i]]@strand
  seq_vec[i] <- motif1@alignList[[i]]@seq
}

out = list(seq_vec=seq_vec, strand_vec = strand_vec, seqID_vec = seqID_vec)
return(out)
}
strand1 = summary_motif(motif11)$strand_vec
strand2 = summary_motif(motif21)$strand_vec
strand3 = summary_motif(motif31)$strand_vec

unq_seqID <- summary_motif(motif11)$seqID_vec %>% unique()

use_id <- match( unq_seqID, summary_motif(motif11)$seqID_vec ) 

use_id_2 <- match(summary_motif(motif21)$seqID_vec %>% unique(),summary_motif(motif21)$seqID_vec )

use_id_3 <- match(summary_motif(motif31)$seqID_vec %>% unique(),summary_motif(motif31)$seqID_vec )
```

Table 1: The proportion of CTCF sequences with different directions (sense or antisense direction) in each of the three clusters. 

```{r}
rrr <- rbind(table(strand1[use_id])/length(strand1[use_id]),
      table(strand2[use_id_2])/(length(strand2[use_id_2])),
      table(strand3[use_id_3])/length(strand3[use_id_3]))

rownames(rrr) <- c("Gained", "Lost", "Stable")
pander::pander(rrr)

```

```{r, eval = F}
rrr <- rbind(table(strand1)/length(strand1),
      table(strand2)[-1]/(length(strand2)-2),
      table(strand3)/length(strand3))

rownames(rrr) <- c("Gained", "Lost", "Stable")
pander::pander(rrr)
```

# Comparing the identified CTCF-like consensus in the three cluster, in a **11 bp** window




```{r}
rr1 = summary_motif(motif11)
rr2 = summary_motif(motif21)
rr3 = summary_motif(motif31)
res =consensusMatrix(DNAStringSet(rr1$seq_vec), as.prob = TRUE)[1:4,]

# I need to transform the seq_vec to its complement

res_revers =consensusMatrix(reverseComplement(DNAStringSet(rr1$seq_vec)), as.prob = TRUE)[1:4,]
#seqLogo(res)
#seqLogo(res_revers) # The center is on position 8

res1 = res_revers[,2:14]
res_melt1<-melt(res1)
colnames(res_melt1)<- c("Nucleotide", "Position", "Freq.")

res =consensusMatrix(DNAStringSet(rr2$seq_vec ), as.prob = TRUE)[1:4,]
res_revers2 =consensusMatrix( reverseComplement(DNAStringSet(rr2$seq_vec)), as.prob = TRUE)[1:4,]

#seqLogo(res)
#seqLogo(res_revers2) # The center is on position 8

res2 = res_revers2 [,2:14]
res_melt2<-melt(res2)
colnames(res_melt2)<- c("Nucleotide", "Position", "Freq.")

res =consensusMatrix(DNAStringSet(rr3$seq_vec ), as.prob = TRUE)[1:4,]
#seqLogo(res) # The center is on position 8
res3 = res[,2:14]
res_melt3<-melt(res3)
colnames(res_melt3)<- c("Nucleotide", "Position", "Freq.")


```



```{r, fig.width = 6, fig.height =3}

#seqLogo(res_revers) 
seqLogo(res1)
#seqLogo(motif11@pwm)
#mtext("Gained Cluster", 3)
#Fig 2: sequence logo of the CTCF motif identifed in the **Gained** Cluster
```

Fig 3: sequence logo of the CTCF motif identifed in the **Gained** Cluster


```{r, fig.width = 6, fig.height =3}
#seqLogo(motif31@pwm[, 3:13])
seqLogo(res3)
```

Fig 4: sequence logo of the CTCF motif identifed in the **Stable** Cluster

```{r, fig.width = 6, fig.height =3}
seqLogo(res2)
#seqLogo(motif21@pwm[,1:11])
#
```

Fig 5: sequence logo of the CTCF motif identifed in the **Loss** Cluster



# Matching Status

In this section, I will look into the identified CTCF-like censensus in the *Gained*, *Lost*, and *Stable* cluster and compare their binding site orientation, and their nucleotide distribution in a longer window (21bp) centered at the mid-point of the canonical CTCF motif, i.e the 10th nucleotide ('G') in the CTCF logo in Fig. 1.

```{r}
# define the GRange object
# Object out.ranges stores all the genomic Ranges -- a total of 78601 ranges
out.ranges<- readRDS("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/WitcherKI/Results/out_ranges_June_28.rds")
out_order <- out.ranges[order(out.ranges$PValue),]

# The top 100 down/up sequences

out_down <-  out_order[out_order$direction =="down",]
out_up <-  out_order[out_order$direction =="up",]

gr_up <- out_up[1:1000,]
gr_down <- out_down[1:1000, ]
gr_null <- out_order[(length(out_order)-999):length(out_order),]




```

```{r}

Seqs_up=Biostrings::getSeq(Hsapiens, GRanges(gr_up) + 120 )
Seqs_down=Biostrings::getSeq(Hsapiens, GRanges(gr_down) + 120 )
Seqs_null=Biostrings::getSeq(Hsapiens, GRanges(gr_null) +120 )  

# in the paper "CTCF maintains regulatory homeostasis of cancer pathways", authors extracted the 500 bp DNA sequences centered at the mid-point of the regions defined in the csaw analysis

# Q: do we need to grow the sequence up to 500 bps?

# Yes (because, if we want to extend the sequence plus 20bp in both sides, some might encounter the boundary)

#see1 = GRanges(seqs_up) + 60
```


```{r, eval=F}

#rr1$seq_vec
res_revers =consensusMatrix(reverseComplement(DNAStringSet(rr1$seq_vec)), as.prob = TRUE)[1:4,]
seqLogo(res_revers)

```


```{r a longer seq_set_extract function}
seq_set_extract <- function(rr1, motif11, Seqs_up,bp.use = 21, reverse = FALSE, rrWidth=13, start = 2, end = 14){
  ext = (bp.use-1)/2 - (rrWidth-1)/2
  #if(width(rr1$seq_vec) %>% unique()==13){  # up sequences identified by rGADEM
  
  #}
  #if(width(rr1$seq_vec) %>% unique() ==18){ # down sequences
  #  start = 1; end = 11; center = 6
  #}
   #if(width(rr1$seq_vec) %>% unique() ==14){ # null sequences
  #  start = 3; end = 13; center = 8
  #}
  seqs_set <- rep(NA, length(rr1$strand_vec))
  
  for ( i in 1:length(rr1$strand_vec)){
    if(rr1$strand_vec[i] == "+"){
      seq_nn <- Seqs_up[motif11@alignList[[i]]@seqID] #the raw input long sequence
  #rr1$seq_vec[i]
oo = gregexpr(DNAString( rr1$seq_vec[i]), seq_nn)

seqs_set[i]<- subseq(seq_nn , start= oo[[1]][1] + (start-1) - ext, end = oo[[1]][1]+(end-1) + ext )
      
     # id_now = rr1$seqID_vec[i]
   #seqs_set[i]<- subseq(Seqs_up[id_now],
    ##   start= motif11@alignList[[i]]@pos+(start-1) - ext ,
     #  motif11@alignList[[i]]@pos+(end-1) + ext  )
    }
    if(rr1$strand_vec[i] == "-"){
      seq_nn <- Seqs_up[motif11@alignList[[i]]@seqID] #the raw input long sequence -- this is in the reverse strand

#oo = gregexpr(DNAString( rr1$seq_vec[i])%>% reverseComplement(), seq_nn)
#if(oo[[1]][1] != -1){

#seqs_set[i]<-subseq(seq_nn %>% reverseComplement(), start= width(seq_nn)-oo[[1]][1]+1-10 -ext-(width(rr1$seq_vec[i])-end) , #width(seq_nn)-oo[[1]][1]+1 + ext -(width(rr1$seq_vec[i])-end) )


# Modify the reverse strand extraction-- July 1 reverse the original input strand instead --> advantage: less tuning for start and end for the - strand

oo = gregexpr(DNAString(rr1$seq_vec[i]), seq_nn %>% reverseComplement())
if(oo[[1]][1] != -1){

seqs_set[i]<- subseq(seq_nn %>% reverseComplement() , start= oo[[1]][1] + (start-1) - ext, end = oo[[1]][1]+(end-1) + ext )



# if(width(rr1$seq_vec) %>% unique()==18){
   
 #  subseq(seq_nn , start= oo[[1]][1] + (start-1) - ext, end = oo[[1]][1]+(end-1) + ext ) %>% reverseComplement()
#seqs_set[i]<-subseq(seq_nn %>% reverseComplement(), start= width(seq_nn)-oo[[1]][1], width(seq_nn)-oo[[1]][1]+1+11 )
#}

}
    }
    
  }
  
  
  #seqs_set <- seqs_set[!is.na(seqs_set)]
  
  if(reverse){
    res =consensusMatrix(reverseComplement(DNAStringSet(seqs_set[!is.na(seqs_set)] )), as.prob = TRUE)[1:4,]
    seqs_set[!is.na(seqs_set)] <- reverseComplement(DNAStringSet(seqs_set[!is.na(seqs_set)] ))
    
  }else{
res =consensusMatrix(DNAStringSet(seqs_set[!is.na(seqs_set)] ), as.prob = TRUE)[1:4,]
  }
  res_melt1<-melt(res)
colnames(res_melt1)<- c("Nucleotide", "Position", "Freq.")
return(list(res = res, res_melt = res_melt1, seqs_set=seqs_set)) 
# April 5, 2021, extracting the raw sequence matrix as well
}
```


```{r}

# there is one sequence in the down cluster that cannot match and provide extended sequences.
res_ext1 <- seq_set_extract(rr1, motif11, Seqs_up, bp.use = 61, reverse = TRUE, rrWidth = 13, start = 1, end = 13)
#-------
# for the up cluster -- the orignal rGADEM identified sequences of length 1-14, center is on position 8
#----------------

# start: start in the original subsequences rr1$seq_vec[i]
# end: end in the original subsequences rr1$seq_vec[i]
# make everything symetric; facilitate any follow-up manipulation for the reverse strand

# For the reverse = TRUE sequences, carefully select the start and end

res_ext2 <- seq_set_extract(rr2, motif21, Seqs_down, bp.use = 61, reverse = TRUE, rrWidth = 13, start = 3, end = 15)

# Down cluster -- 16 bp, center is on position 8
# For the reverse strend, the point is to specify the end position

res_ext3<- seq_set_extract(rr3, motif31, Seqs_null, bp.use = 61, reverse = FALSE,
                           rrWidth = 13, start = 2, end = 14)
```



```{r}
seqs_set1 <- res_ext1$seqs_set  # A sequence of length 61
seqs_set2 <- res_ext2$seqs_set
seqs_set3 <- res_ext3$seqs_set


# Check

seqLogo( consensusMatrix(DNAStringSet(seqs_set1), as.prob = TRUE)[1:4,])
seqLogo( consensusMatrix(DNAStringSet(seqs_set2[!is.na(seqs_set2)]), as.prob = TRUE)[1:4,])
seqLogo( consensusMatrix(DNAStringSet(seqs_set3), as.prob = TRUE)[1:4,])

# Matching this to the original input
matchID1 <- rr1$seqID_vec
matchID2 <- rr2$seqID_vec
matchID3 <- rr3$seqID_vec

seqs_set1_sub <- seqs_set1
seqs_set2_sub <- seqs_set2
seqs_set3_sub <- seqs_set3



for( i in 1:length(seqs_set1)){
  seqs_set1_sub[i] <- subseq(seqs_set1[i], 25, 48)
}
for( i in which(!is.na(seqs_set2))){
  seqs_set2_sub[i] <- subseq(seqs_set2[i], 25, 48)
}
for( i in 1:length(seqs_set3)){
  seqs_set3_sub[i] <- subseq(seqs_set3[i], 25, 48)
}

# Check

seqLogo( consensusMatrix(DNAStringSet(seqs_set1_sub), as.prob = TRUE)[1:4,])
seqLogo( consensusMatrix(DNAStringSet(seqs_set2_sub[!is.na(seqs_set2_sub)]), as.prob = TRUE)[1:4,])
seqLogo( consensusMatrix(DNAStringSet(seqs_set3_sub), as.prob = TRUE)[1:4,])


df1 <- as.data.frame(gr_up)
df2 <- as.data.frame(gr_down)
df3 <- as.data.frame(gr_null)


df1$ctcf_like <- 0
df2$ctcf_like <- 0
df3$ctcf_like <- 0

df1$ctcf_like[matchID1] <- 1
df2$ctcf_like[matchID2] <- 1
df3$ctcf_like[matchID3] <- 1

df1$sequence <- NA
df2$sequence <- NA
df3$sequence <- NA

unq_seqID <- matchID1 %>% unique()

use_id <- match( unq_seqID, matchID1 ) 

use_id_2 <- match(matchID2 %>% unique(), matchID2)

use_id_3 <- match(matchID3 %>% unique(), matchID3)

df1$sequence[unique(matchID1)] <- seqs_set1_sub[use_id]
df2$sequence[unique(matchID2)] <- seqs_set2_sub[use_id_2]
df3$sequence[unique(matchID3)] <- seqs_set3_sub[use_id_3]


df1$Pos16[which(df1$ctcf_like==1)] <- 
vapply(which(df1$ctcf_like==1), function(i){
  subseq((df1$sequence)[i], 16, 16)
}, "A")

df1$Pos19[which(df1$ctcf_like==1)] <- 
vapply(which(df1$ctcf_like==1), function(i){
  try({subseq((df1$sequence)[i], 19, 19)})
}, "A")
df1$Pos22[which(df1$ctcf_like==1)] <- 
vapply(which(df1$ctcf_like==1), function(i){
  try({subseq((df1$sequence)[i], 22, 22)})
}, "A")


seqLogo(consensusMatrix(DNAStringSet(df2$sequence[!is.na(df2$sequence)]), as.prob = TRUE)[1:4,])

df2$Pos16 <- df2$Pos19 <- df2$Pos22 <- NA
df2$Pos16[which(df2$ctcf_like==1)] <- 
vapply(which(df2$ctcf_like==1), function(i){
  subseq((df2$sequence)[i], 16, 16)
}, "A")

df2$Pos19[which(df2$ctcf_like==1)] <- 
vapply(which(df2$ctcf_like==1), function(i){
  subseq((df2$sequence)[i], 19, 19)
}, "A")
df2$Pos22[which(df2$ctcf_like==1)] <- 
vapply(which(df2$ctcf_like==1), function(i){
  subseq((df2$sequence)[i], 22, 22)
}, "A")



for( i in which(df3$ctcf_like==1)){
  df3$Pos16[i] = subseq((df3$sequence)[i], 16, 16)
  df3$Pos19[i] = subseq((df3$sequence)[i], 19, 19)
  df3$Pos22[i] = subseq((df3$sequence)[i], 22, 22)
}



head(df3)
```



```{r}
df1$isPos16A <- df1$Pos16=="A"
df1$isPos19G <- df1$Pos19=="G"
df1$isPos22C <- df1$Pos22=="C"

df2$isPos16A <- df2$Pos16=="A"
df2$isPos19G <- df2$Pos19=="G"
df2$isPos22C <- df2$Pos22=="C"

df3$isPos16A <- df3$Pos16=="A"
df3$isPos19G <- df3$Pos19=="G"
df3$isPos22C <- df3$Pos22=="C"
```





```{r}
logo = res_ext2$res[,25:48]

w1 =logo["A",16]
w2 = logo["G",19]
w3 =logo["C",22]


df1$isAGC <- df1$isPos16A * w1 + df1$isPos19G *w2 + df1$isPos22C* w3
df2$isAGC <- df2$isPos16A * w1 + df2$isPos19G *w2 + df2$isPos22C* w3
df3$isAGC <- df3$isPos16A * w1 + df3$isPos19G *w2 + df3$isPos22C* w3

# df1
score_16_24 <- rep(NA, nrow(df1))
for( i in which(df1$ctcf_like==1)){
  ss <- 0
  for( pp in 16:24){
    ll =  subseq((df1$sequence)[i], pp, pp) 
   ss = ss + logo[ll, pp]
  }
  score_16_24[i]=ss 
}
df1$score_16_24 = score_16_24

# df2
score_16_24 <- rep(NA, nrow(df2))
for( i in which(df2$ctcf_like==1)){
  ss <- 0
  for( pp in 16:24){
    ll =  subseq((df2$sequence)[i], pp, pp) 
   ss = ss + logo[ll, pp]
  }
  score_16_24[i]=ss 
}
df2$score_16_24 = score_16_24

# df3
score_16_24 <- rep(NA, nrow(df3))
for( i in which(df3$ctcf_like==1)){
  ss <- 0
  for( pp in 16:24){
    ll =  subseq((df3$sequence)[i], pp, pp) 
   ss = ss + logo[ll, pp]
  }
  score_16_24[i]=ss 
}
df3$score_16_24 = score_16_24
#logo[c("A", "G", "C"),c(16, 19, 22)]



df1$score_16_24Greater2.5 <- df1$score_16_24>2.5
df2$score_16_24Greater2.5 <- df2$score_16_24>2.5
df3$score_16_24Greater2.5 <- df3$score_16_24>2.5


df1$isAGCGreater0.75 <- df1$isAGC >0.75
df2$isAGCGreater0.75 <- df2$isAGC >0.75
df3$isAGCGreater0.75 <- df3$isAGC >0.75
```

```{r}
setwd("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/WitcherKI/4-extract-matching-status")
write.csv(df1, file = "Gain.csv")
write.csv(df2, file = "Lost.csv")
write.csv(df3, file = "Stable.csv")
```



```{r}
boxplot(data.frame(df1$score_16_24,  df3$score_16_24, df2$score_16_24), names= 
          c("Gained", "Stable", "Lost"), pch = 20, main = "Score based on seq 16:24")

table(df1$isPos16A[df1$ctcf_like==1])/(sum(df1$ctcf_like==1))
table(df2$isPos16A[df2$ctcf_like==1])/(sum(df2$ctcf_like==1))
table(df3$isPos16A[df3$ctcf_like==1])/(sum(df3$ctcf_like==1))

table(df1$isPos19G[df1$ctcf_like==1])/(sum(df1$ctcf_like==1))
table(df2$isPos19G[df2$ctcf_like==1])/(sum(df2$ctcf_like==1))
table(df3$isPos19G[df3$ctcf_like==1])/(sum(df3$ctcf_like==1))


table(df1$isPos22C[df1$ctcf_like==1])/(sum(df1$ctcf_like==1))
table(df2$isPos22C[df2$ctcf_like==1])/(sum(df2$ctcf_like==1))
table(df3$isPos22C[df3$ctcf_like==1])/(sum(df3$ctcf_like==1))

```

I generate the dataframe with small segments indicated by 

Reformat the 1000 sequences into small segments obtained from rGADEM 

```{r, eval = F}

Reformat_dfs <- function(rr1, seqs_set1, motif11 ){
  matchID1 <- rr1$seqID_vec
  
  seqs_set1_sub <- seqs_set1
  for( i in 1:length(seqs_set1)){
    seqs_set1_sub[i] <- subseq(seqs_set1[i], 25, 48)
  }
  
  dfsee = NULL
  
  for( i in seq(length(matchID1))){
    
    # df1[matchID1[i],][]
    now_segs <- which(matchID1 == matchID1[i])
    
    for( jj in now_segs){
      cur_seq <- motif11@alignList[[jj]]
      
      dfsee <- rbind( dfsee, c(cur_seq@chr, cur_seq@start, cur_seq@end, seqs_set1_sub[jj] , matchID1[jj]))
    }
    
  }
  
  colnames(dfsee) <- c("chr", "start", "end", "sequences", "seqID_to_1000subset")
  
  return(dfsee)
}
  
df_expand_1 <- Reformat_dfs(rr1, seqs_set1, motif11)
df_expand_2 <- Reformat_dfs(rr2, seqs_set2, motif21)
df_expand_3 <- Reformat_dfs(rr3, seqs_set3, motif31)




```



```{r, eval=TRUE}
knitr::knit_exit()
```
