---
title: "Comparison of CTCF consensus found in the (mutation-induced) Gained, Lost and stable binding sites: a descriptive analysis"
author: "by Kaiqiong Zhao"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: false
    number_sections: true
    toc_depth: 4
    keep_md: true
editor_options: 
  chunk_output_type: console
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, fig.width = 10, fig.height = 4) 
```



```{r}
library(magrittr)
library(rGADEM)
library(ggplot2)
library(reshape2)
library(MotIV)
library(BSgenome.Hsapiens.UCSC.hg19)
```

# Overview

The previous report, "(2)Motif_discovery_rGADEM.html", has shown that the most significant motif indentified in each of the three clusters, **Gained**, **Lost** and **Stable**, is the canonical CTCF motif. In this report, I will look into these identified CTCF-like censensus in each cluster and compare their binding site orientation, and their nucleotide distribution in windows with different lengths centered at the mid-point of the canonical CTCF motif, i.e the 10th nucleotide ('G') in the CTCF logo.

```{r, fig.width=10, fig.h=8}
path<- "/scratch/greenwood/kaiqiong.zhao/Old_home/kaiqiong.zhao/R/x86_64-redhat-linux-gnu-library/3.5/rGADEM/extdata/jaspar2009.txt"

seededPwm<-readPWMfile(path)
#grep("STAT1",names(seededPwm))

spwm = seededPwm$CTCF
seqLogo(apply(spwm, 2, function(x){x/sum(x)}))
```

Fig 1: sequence logo of the canonical CTCF motif


```{r}
#load("~/scratch/kaiqiong.zhao/Projects/Maika/Maika_Chipseq/Report/GADEM_up.RData")
#load("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/WitcherKI/2-GADEM_all_cluster/GADEM_up.RData")
#gadem_up <- gadem

load("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/WitcherKI/2-GADEM_all_cluster/GADEM_down.RData")

gadem_down <- gadem

#load("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/Maika/Maika_Chipseq/Descrip-aug/GADEM_null_top1000.RData")
#load("~/scratch/kaiqiong.zhao/Projects/Maika/Maika_Chipseq/Descrip-aug/GADEM_null_top1000.RData")
#gadem_null <- gadem

#motif11 <- gadem_up@motifList[[1]]
motif21 <- gadem_down@motifList[[1]]
#motif31 <- gadem_null@motifList[[1]]
```

# CTCF binding site orientations in the three cluster

```{r}

summary_motif <- function(motif1){
seqID_vec <- rep(NA, length(motif1@alignList))
seq_vec <- strand_vec <- pval <- seqID_vec
 
for ( i in 1:length(seqID_vec)){
  seqID_vec[i] <- motif1@alignList[[i]]@seqID
  pval[i] <- motif1@alignList[[i]]@pval
  strand_vec[i] <- motif1@alignList[[i]]@strand
  seq_vec[i] <- motif1@alignList[[i]]@seq
}

out = list(seq_vec=seq_vec, strand_vec = strand_vec, seqID_vec = seqID_vec)
return(out)
}
#strand1 = summary_motif(motif11)$strand_vec
strand2 = summary_motif(motif21)$strand_vec
#strand3 = summary_motif(motif31)$strand_vec

#unq_seqID <- summary_motif(motif11)$seqID_vec %>% unique()

#use_id <- match( unq_seqID, summary_motif(motif11)$seqID_vec ) 

use_id_2 <- match(summary_motif(motif21)$seqID_vec %>% unique(),summary_motif(motif21)$seqID_vec )

#use_id_3 <- match(summary_motif(motif31)$seqID_vec %>% unique(),summary_motif(motif31)$seqID_vec )
```


# Comparing the identified CTCF-like consensus in the three cluster, in a **11 bp** window




```{r}
#rr1 = summary_motif(motif11)
rr2 = summary_motif(motif21)
#rr3 = summary_motif(motif31)

res =consensusMatrix(DNAStringSet(rr2$seq_vec ), as.prob = TRUE)[1:4,]
res_revers2 =consensusMatrix( reverseComplement(DNAStringSet(rr2$seq_vec)), as.prob = TRUE)[1:4,]

seqLogo(res_revers2)
res2 = res_revers2[,1:13]


res_melt2<-melt(res2)
colnames(res_melt2)<- c("Nucleotide", "Position", "Freq.")



```





```{r, fig.width = 6, fig.height =3}

seqLogo(res2)
#
```

Fig 5: sequence logo of the CTCF motif identifed in the **Loss** Cluster



# Comparing the identified CTCF-like consensus in the three cluster, in a **61 bp** window

In this section, I will look into the identified CTCF-like censensus in the *Gained*, *Lost*, and *Stable* cluster and compare their binding site orientation, and their nucleotide distribution in a longer window (21bp) centered at the mid-point of the canonical CTCF motif, i.e the 10th nucleotide ('G') in the CTCF logo in Fig. 1.

```{r}

#-----------------------------
# Make sure the orders are the same as the input in "Report/Run-GADEM-negative.R"
#--------------------------

out.ranges<- readRDS("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/WitcherKI/Results/out_ranges_June_28.rds")
Sig_res <- out.ranges[out.ranges$FDR <=0.05,]  #24714 ranges 

#Sig_res_up <- Sig_res[Sig_res$direction=="up",] # 10567

gr_down <- Sig_res[Sig_res$direction=="down",] # 14136


#out_order_sig <- (out_order[out_order$FDR<0.05,])

#gr_up  <-(out_order_sig[out_order_sig$direction=="up" ,])

#gr_down <-(out_order_sig[out_order_sig$direction=="down" ,])

#gr_null <- (out_order[out_order$FDR>=0.05,])


# convert GRanges back to a RangedData
#seqs_up<-RangedData(gr_up@ranges,space = gr_up@seqnames)
#seqs_down<-RangedData(gr_down@ranges,space = gr_down@seqnames)
#seqs_null<-RangedData(gr_null@ranges,space = gr_null@seqnames)

```




```{r}

#Seqs_up=Biostrings::getSeq(Hsapiens, GRanges(seqs_up) + 60 )
Seqs_down=Biostrings::getSeq(Hsapiens, GRanges(gr_down) + 160 )
#Seqs_null=Biostrings::getSeq(Hsapiens, GRanges(seqs_null) +60 )  

# in the paper "CTCF maintains regulatory homeostasis of cancer pathways", authors extracted the 500 bp DNA sequences centered at the mid-point of the regions defined in the csaw analysis

# Q: do we need to grow the sequence up to 500 bps?

# Yes (because, if we want to extend the sequence plus 20bp in both sides, some might encounter the boundary)

#see1 = GRanges(seqs_up) + 60

```





```{r}
# define a function to extract a set of sequences of length bp.use and calculate the Nucleotide probability

seq_set_extract <- function(rr1, motif11, Seqs_up,bp.use = 21, reverse = FALSE, rrWidth=13, start = 2, end = 14){
  ext = (bp.use-1)/2 - (rrWidth-1)/2
  #if(width(rr1$seq_vec) %>% unique()==13){  # up sequences identified by rGADEM
  
  #}
  #if(width(rr1$seq_vec) %>% unique() ==18){ # down sequences
  #  start = 1; end = 11; center = 6
  #}
   #if(width(rr1$seq_vec) %>% unique() ==14){ # null sequences
  #  start = 3; end = 13; center = 8
  #}
  seqs_set <- rep(NA, length(rr1$strand_vec))
  
  for ( i in 1:length(rr1$strand_vec)){
    if(rr1$strand_vec[i] == "+"){
      seq_nn <- Seqs_up[motif11@alignList[[i]]@seqID] #the raw input long sequence
  #rr1$seq_vec[i]
oo = gregexpr(DNAString( rr1$seq_vec[i]), seq_nn)

seqs_set[i]<- subseq(seq_nn , start= oo[[1]][1] + (start-1) - ext, end = oo[[1]][1]+(end-1) + ext )
      
     # id_now = rr1$seqID_vec[i]
   #seqs_set[i]<- subseq(Seqs_up[id_now],
    ##   start= motif11@alignList[[i]]@pos+(start-1) - ext ,
     #  motif11@alignList[[i]]@pos+(end-1) + ext  )
    }
    if(rr1$strand_vec[i] == "-"){
      seq_nn <- Seqs_up[motif11@alignList[[i]]@seqID] #the raw input long sequence -- this is in the reverse strand

#oo = gregexpr(DNAString( rr1$seq_vec[i])%>% reverseComplement(), seq_nn)
#if(oo[[1]][1] != -1){

#seqs_set[i]<-subseq(seq_nn %>% reverseComplement(), start= width(seq_nn)-oo[[1]][1]+1-10 -ext-(width(rr1$seq_vec[i])-end) , #width(seq_nn)-oo[[1]][1]+1 + ext -(width(rr1$seq_vec[i])-end) )


# Modify the reverse strand extraction-- July 1 reverse the original input strand instead --> advantage: less tuning for start and end for the - strand

oo = gregexpr(DNAString(rr1$seq_vec[i]), seq_nn %>% reverseComplement())
if(oo[[1]][1] != -1){

seqs_set[i]<- subseq(seq_nn %>% reverseComplement() , start= oo[[1]][1] + (start-1) - ext, end = oo[[1]][1]+(end-1) + ext )



# if(width(rr1$seq_vec) %>% unique()==18){
   
 #  subseq(seq_nn , start= oo[[1]][1] + (start-1) - ext, end = oo[[1]][1]+(end-1) + ext ) %>% reverseComplement()
#seqs_set[i]<-subseq(seq_nn %>% reverseComplement(), start= width(seq_nn)-oo[[1]][1], width(seq_nn)-oo[[1]][1]+1+11 )
#}

}
    }
    
  }
  
  
  #seqs_set <- seqs_set[!is.na(seqs_set)]
  
  if(reverse){
    res =consensusMatrix(reverseComplement(DNAStringSet(seqs_set[!is.na(seqs_set)] )), as.prob = TRUE)[1:4,]
    seqs_set[!is.na(seqs_set)] <- reverseComplement(DNAStringSet(seqs_set[!is.na(seqs_set)] ))
    
  }else{
res =consensusMatrix(DNAStringSet(seqs_set[!is.na(seqs_set)] ), as.prob = TRUE)[1:4,]
  }
  res_melt1<-melt(res)
colnames(res_melt1)<- c("Nucleotide", "Position", "Freq.")
return(list(res = res, res_melt = res_melt1, seqs_set=seqs_set)) 
# April 5, 2021, extracting the raw sequence matrix as well
}

```



```{r}
# in the cluster "Down", there are 2 sequences without the strand information

seqLogo(res_revers2)
res_ext2 <- seq_set_extract(rr2, motif21, Seqs_down, bp.use = 61, reverse = TRUE, rrWidth = 13, start = 7, end = 19)

```





# Matching status




```{r}
#seqs_set1 <- res_ext1$seqs_set  # A sequence of length 61
seqs_set2 <- res_ext2$seqs_set
#seqs_set3 <- res_ext3$seqs_set


seqLogo( consensusMatrix(DNAStringSet(seqs_set2[!is.na(seqs_set2)]), as.prob = TRUE)[1:4,])

# Matching this to the original input
#matchID1 <- rr1$seqID_vec
matchID2 <- rr2$seqID_vec
#matchID3 <- rr3$seqID_vec

#seqs_set1_sub <- seqs_set1
seqs_set2_sub <- seqs_set2
#seqs_set3_sub <- seqs_set3




for( i in which(!is.na(seqs_set2))){
  seqs_set2_sub[i] <- subseq(seqs_set2[i], 25, 48)
}


seqLogo( consensusMatrix(DNAStringSet(seqs_set2_sub[!is.na(seqs_set2_sub)]), as.prob = TRUE)[1:4,])

df2 <- as.data.frame(gr_down)



#df1$ctcf_like <- 0
df2$ctcf_like <- 0
#df3$ctcf_like <- 0

#df1$ctcf_like[matchID1] <- 1
df2$ctcf_like[matchID2] <- 1
#df3$ctcf_like[matchID3] <- 1

#df1$sequence <- NA
df2$sequence <- NA
#df3$sequence <- NA

#unq_seqID <- matchID1 %>% unique()

#use_id <- match( unq_seqID, matchID1 ) 

use_id_2 <- match(matchID2 %>% unique(), matchID2)

#use_id_3 <- match(matchID3 %>% unique(), matchID3)

#df1$sequence[unique(matchID1)] <- seqs_set1_sub[use_id]
df2$sequence[unique(matchID2)] <- seqs_set2_sub[use_id_2]
#df3$sequence[unique(matchID3)] <- seqs_set3_sub[use_id_3]


# if multiple subsequences match to the same line in df2, use the first one





df2$Pos16 <- df2$Pos19 <- df2$Pos22 <- rep(NA, nrow(df2))

for( i in which(df2$ctcf_like==1)){
  try({
  df2$Pos16[i] = subseq((df2$sequence)[i], 16, 16)
  df2$Pos19[i] = subseq((df2$sequence)[i], 19, 19)
  df2$Pos22[i] = subseq((df2$sequence)[i], 22, 22)
  })
}



```



```{r}

df2$isPos16A <- df2$Pos16=="A"
df2$isPos19G <- df2$Pos19=="G"
df2$isPos22C <- df2$Pos22=="C"


```





```{r}
logo = res_ext2$res[,25:48]

w1 =logo["A",16]
w2 = logo["G",19]
w3 =logo["C",22]


#df1$isAGC <- df1$isPos16A * w1 + df1$isPos19G *w2 + df1$isPos22C* w3
df2$isAGC <- df2$isPos16A * w1 + df2$isPos19G *w2 + df2$isPos22C* w3
#df3$isAGC <- df3$isPos16A * w1 + df3$isPos19G *w2 + df3$isPos22C* w3


# df2
score_16_24 <- rep(NA, nrow(df2))
for( i in which(df2$ctcf_like==1)){
  try({
  ss <- 0
  for( pp in 16:24){
    ll =  subseq((df2$sequence)[i], pp, pp) 
   ss = ss + logo[ll, pp]
  }
  score_16_24[i]=ss 
  })
}
df2$score_16_24 = score_16_24



#df1$score_16_24Greater2.5 <- df1$score_16_24>2.5
df2$score_16_24Greater2.5 <- df2$score_16_24>2.5
#df3$score_16_24Greater2.5 <- df3$score_16_24>2.5


#df1$isAGCGreater0.75 <- df1$isAGC >0.75
df2$isAGCGreater0.75 <- df2$isAGC >0.75
#df3$isAGCGreater0.75 <- df3$isAGC >0.75
```

```{r}

setwd("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/WitcherKI/4-extract-matching-status")

write.csv(df2, file = "WholeLost.csv")
```



```{r}
boxplot(data.frame(df1$score_16_24,  df3$score_16_24, df2$score_16_24), names= 
          c("Gained", "Stable", "Lost"), pch = 20, main = "Score based on seq 16:24")

table(df2$isPos16A)
table(df2$isPos19G)
table(df2$isPos22C)
```

I generate the dataframe with small segments indicated by 

Reformat the 1000 sequences into small segments obtained from rGADEM 

```{r, eval = F}

Reformat_dfs <- function(rr1, seqs_set1, motif11 ){
  matchID1 <- rr1$seqID_vec
  
  seqs_set1_sub <- seqs_set1
  for( i in 1:length(seqs_set1)){
    seqs_set1_sub[i] <- subseq(seqs_set1[i], 25, 48)
  }
  
  dfsee = NULL
  
  for( i in seq(length(matchID1))){
    
    # df1[matchID1[i],][]
    now_segs <- which(matchID1 == matchID1[i])
    
    for( jj in now_segs){
      cur_seq <- motif11@alignList[[jj]]
      
      dfsee <- rbind( dfsee, c(cur_seq@chr, cur_seq@start, cur_seq@end, seqs_set1_sub[jj] , matchID1[jj]))
    }
    
  }
  
  colnames(dfsee) <- c("chr", "start", "end", "sequences", "seqID_to_1000subset")
  
  return(dfsee)
}
  
df_expand_1 <- Reformat_dfs(rr1, seqs_set1, motif11)
df_expand_2 <- Reformat_dfs(rr2, seqs_set2, motif21)
df_expand_3 <- Reformat_dfs(rr3, seqs_set3, motif31)




```



```{r}
res1 = res_ext1$res_melt
res2 = res_ext2$res_melt
res3 = res_ext3$res_melt
```


# Comparing the identified CTCF-like consensus in the three cluster, in a **61 bp** window


```{r}

res1_21<- res1
res2_21<- res2
res3_21<- res3


res_tab = cbind(res1_21, res3_21[,3], res2_21[,3])
colnames(res_tab)[3:5] <- c("Gained", "Stable", "Lost")
rownames(res_tab) = NULL
#pander::pander(res_tab)

```


```{r, fig.width=18, fig.height=6}
res_span <- melt(res_tab, id.vars = c("Nucleotide", "Position"))
colnames(res_span)[3:4] <- c("Type", "Freq.")

ggplot(res_span, aes(Position, Freq.))+
  geom_bar(stat = "identity", aes(fill = Nucleotide))+
  geom_vline(xintercept = 31, 
             color = "blue", size=1)+
  facet_wrap(~Type, nrow = 1)
```


Fig 14: Necleotide distribution in a 61bp window centered at the mid-point of the canonical CTCF motif for binding sequences in cluster Gained, stable and lost. The vertical blue line indicates the center which corresponds to the 10th nucleotide ('G') in Fig 1.


```{r, fig.width = 18, fig.height =3}
seqLogo( res_ext1$res)

```


Fig 15: sequence logo of the CTCF motif identifed in the **Gained** Cluster

```{r, fig.width = 18, fig.height =3}

seqLogo( res_ext3$res)


```

Fig 16: sequence logo of the CTCF motif identifed in the **Stable** Cluster

```{r, fig.width = 18, fig.height =3}

seqLogo( res_ext2$res)

```

Fig 17: sequence logo of the CTCF motif identifed in the **Lost** Cluster

# Frequency differences


## **Lost** v.s. **Stable**

```{r}

res_tab$`lost-stable` <- res_tab$Lost - res_tab$Stable
res_tab$`Gained-stable` <- res_tab$Gained - res_tab$Stable

ggplot(res_tab, aes(Position, `lost-stable`))+
  geom_point(stat = "identity" , aes(colour = Nucleotide, shape = Nucleotide))+
  geom_vline(xintercept = 31, 
             color = "blue", size=0.5, linetype = 3)+
  geom_vline(xintercept = 25, linetype = 2, size=0.5) +
    geom_vline(xintercept = 48, linetype = 2, size=0.5) +
  ggtitle("Lost cluster v.s Stable Cluster")+
  ylab("Freq. differences")
```

Fig 18: Difference of nucleotide differences between the **Lost** and **Stable** clusters. Black dashed lines (at position 25 and 48) are the boundaries of the region with frequency differences greater than 10% between the Lost cluster and the stable cluster

Table: Nucleotide frequences at positions with Freq. differences greater than 10% between the **Lost** and **Stable** clusters.

```{r}
see = order(abs(res_tab$`lost-stable`), decreasing = T)
res_tab_ord = res_tab[see,]
ss = res_tab_ord
rownames(ss) =NULL
ss = ss[abs(ss$`lost-stable`)>0.1,]
pander::pander(ss )
```




## **Gained** v.s. **Stable**

```{r}
ggplot(res_tab, aes(Position, `Gained-stable`))+
  geom_point(stat = "identity" , aes(colour = Nucleotide, shape = Nucleotide))+
 geom_vline(xintercept = 31, 
             color = "blue", size=0.5, linetype = 3)+
  geom_vline(xintercept = 33, linetype = 2, size=0.5) +
    geom_vline(xintercept = 47, linetype = 2, size=0.5) +
  ggtitle("Gained cluster v.s Stable Cluster")+
  ylab("Freq. differences")+
  coord_cartesian(ylim = c(-0.2, 0.45))
```

Fig 19: Difference of nucleotide differences between the **Lost** and **Stable** clusters. Black dashed lines (at position 33 and 47) are the boundaries of the region with frequency differences greater than 10% between the Gained cluster and the Stable cluster.

Table: Nucleotide frequences at positions with Freq. differences greater than 10% between the Gained cluster and the Stable cluster.


```{r}
see = order(abs(res_tab$`Gained-stable`), decreasing = T)
res_tab_ord = res_tab[see,]
ss = res_tab_ord
rownames(ss) =NULL
ss = ss[abs(ss$`Gained-stable`)>0.1,]
pander::pander(ss )

```

## Nucleotide frequencies in a **24 bp** window (this region has Freq. differences greater than 10%)




```{r}

res1_21<- res1[res1$Position  %in% 25:48,] # 21bp
res2_21<- res2[res2$Position  %in% 25:48,] # 21bp
res3_21<- res3[res3$Position  %in% 25:48,] # 21bp

res1_21$Position = res1_21$Position-24
res2_21$Position = res2_21$Position-24
res3_21$Position = res3_21$Position-24



res_tab = cbind(res1_21, res3_21[,3], res2_21[,3])
colnames(res_tab)[3:5] <- c("Gained", "Stable", "Lost")
rownames(res_tab) = NULL
#pander::pander(res_tab)

```


```{r, fig.width=12, fig.height=4}
res_span <- melt(res_tab, id.vars = c("Nucleotide", "Position"))
colnames(res_span)[3:4] <- c("Type", "Freq.")

ggplot(res_span, aes(Position, Freq.))+
  geom_bar(stat = "identity", aes(fill = Nucleotide))+
  geom_vline(xintercept = 31-24, 
             color = "blue", size=1)+
  facet_wrap(~Type, nrow = 1)
```


Fig 20: Necleotide distribution in a window for binding sequences in cluster Gained, stable and lost. The vertical blue line indicates the center which corresponds to the 10th nucleotide ('G') in Fig 1.


```{r, fig.width = 18, fig.height =3}
seqLogo( res_ext1$res[,25:48])

```


Fig 21: sequence logo of the CTCF motif identifed in the **Gained** Cluster

```{r, fig.width = 18, fig.height =3}

seqLogo( res_ext3$res[,25:48])


```

Fig 22: sequence logo of the CTCF motif identifed in the **Stable** Cluster

```{r, fig.width = 18, fig.height =3}

seqLogo( res_ext2$res[,25:48])

```

Fig 23: sequence logo of the CTCF motif identifed in the **Lost** Cluster



```{r, eval = F}
library(compositions)

?acomp

sss = acomp(t(res_ext1$res)[26:36,])
plot(sss, pch = 19)
```



```{r, eval=TRUE}
knitr::knit_exit()
```
