---
title: "Comparison of CTCF consensus found in the (mutation-induced) Gained, Lost and stable binding sites: a descriptive analysis"
author: "by Kaiqiong Zhao"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: false
    number_sections: true
    toc_depth: 4
    keep_md: true
editor_options: 
  chunk_output_type: console
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, fig.width = 10, fig.height = 4) 
```



```{r}
library(magrittr)
library(rGADEM)
library(ggplot2)
library(reshape2)
library(MotIV)
library(BSgenome.Hsapiens.UCSC.hg19)
```

# Overview

The previous report, "(2)Motif_discovery_rGADEM.html", has shown that the most significant motif indentified in each of the three clusters, **Gained**, **Lost** and **Stable**, is the canonical CTCF motif. In this report, I will look into these identified CTCF-like censensus in each cluster and compare their binding site orientation, and their nucleotide distribution in windows with different lengths centered at the mid-point of the canonical CTCF motif, i.e the 10th nucleotide ('G') in the CTCF logo.

```{r, fig.width=10, fig.h=8}
path<- "/scratch/greenwood/kaiqiong.zhao/Old_home/kaiqiong.zhao/R/x86_64-redhat-linux-gnu-library/3.5/rGADEM/extdata/jaspar2009.txt"

seededPwm<-readPWMfile(path)
#grep("STAT1",names(seededPwm))

spwm = seededPwm$CTCF
seqLogo(apply(spwm, 2, function(x){x/sum(x)}))
```

Fig 1: sequence logo of the canonical CTCF motif


```{r}
load("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/WitcherKI/2-GADEM_top_1000/GADEM_up_1000.RData")
gadem_up <- gadem
load("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/WitcherKI/2-GADEM_top_1000/GADEM_down_1000.RData")
gadem_down <- gadem
load("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/WitcherKI/2-GADEM_top_1000/GADEM_null_1000.RData")
gadem_null <- gadem

motif11 <- gadem_up@motifList[[1]]
motif21 <- gadem_down@motifList[[1]]
motif31 <- gadem_null@motifList[[1]]
```

# CTCF binding site orientations in the three cluster

```{r}

summary_motif <- function(motif1){
seqID_vec <- rep(NA, length(motif1@alignList))
seq_vec <- strand_vec <- pval <- seqID_vec
 
for ( i in 1:length(seqID_vec)){
  seqID_vec[i] <- motif1@alignList[[i]]@seqID
  pval[i] <- motif1@alignList[[i]]@pval
  strand_vec[i] <- motif1@alignList[[i]]@strand
  seq_vec[i] <- motif1@alignList[[i]]@seq
}

out = list(seq_vec=seq_vec, strand_vec = strand_vec, seqID_vec = seqID_vec)
return(out)
}
strand1 = summary_motif(motif11)$strand_vec
strand2 = summary_motif(motif21)$strand_vec
strand3 = summary_motif(motif31)$strand_vec

unq_seqID <- summary_motif(motif11)$seqID_vec %>% unique()

use_id <- match( unq_seqID, summary_motif(motif11)$seqID_vec ) 

use_id_2 <- match(summary_motif(motif21)$seqID_vec %>% unique(),summary_motif(motif21)$seqID_vec )

use_id_3 <- match(summary_motif(motif31)$seqID_vec %>% unique(),summary_motif(motif31)$seqID_vec )
```

Table 1: The proportion of CTCF sequences with different directions (sense or antisense direction) in each of the three clusters. 

```{r}
rrr <- rbind(table(strand1[use_id])/length(strand1[use_id]),
      table(strand2[use_id_2])/(length(strand2[use_id_2])),
      table(strand3[use_id_3])/length(strand3[use_id_3]))

rownames(rrr) <- c("Gained", "Lost", "Stable")
pander::pander(rrr)

```

```{r, eval = F}
rrr <- rbind(table(strand1)/length(strand1),
      table(strand2)[-1]/(length(strand2)-2),
      table(strand3)/length(strand3))

rownames(rrr) <- c("Gained", "Lost", "Stable")
pander::pander(rrr)
```

# Comparing the identified CTCF-like consensus in the three cluster, in a **11 bp** window




```{r}
rr1 = summary_motif(motif11)
rr2 = summary_motif(motif21)
rr3 = summary_motif(motif31)
res =consensusMatrix(DNAStringSet(rr1$seq_vec), as.prob = TRUE)[1:4,]

# I need to transform the seq_vec to its complement

res_revers =consensusMatrix(reverseComplement(DNAStringSet(rr1$seq_vec)), as.prob = TRUE)[1:4,]
#seqLogo(res)
#seqLogo(res_revers) # The center is on position 8

res1 = res_revers[,2:14]
res_melt1<-melt(res1)
colnames(res_melt1)<- c("Nucleotide", "Position", "Freq.")

res =consensusMatrix(DNAStringSet(rr2$seq_vec ), as.prob = TRUE)[1:4,]
res_revers2 =consensusMatrix( reverseComplement(DNAStringSet(rr2$seq_vec)), as.prob = TRUE)[1:4,]

#seqLogo(res)
#seqLogo(res_revers2) # The center is on position 8

res2 = res_revers2 [,2:14]
res_melt2<-melt(res2)
colnames(res_melt2)<- c("Nucleotide", "Position", "Freq.")

res =consensusMatrix(DNAStringSet(rr3$seq_vec ), as.prob = TRUE)[1:4,]
#seqLogo(res) # The center is on position 8
res3 = res[,2:14]
res_melt3<-melt(res3)
colnames(res_melt3)<- c("Nucleotide", "Position", "Freq.")

library(ggpubr)
library(cowplot)
gg1 = ggplot(res_melt1, aes(Position, Freq.))+
  geom_bar(stat = "identity", aes(fill = Nucleotide))+
  geom_vline(xintercept = 7, 
             color = "blue", size=1)+
  ggtitle("Gained Cluster")
gg2 = ggplot(res_melt2, aes(Position, Freq.))+
  geom_bar(stat = "identity", aes(fill = Nucleotide))+
  geom_vline(xintercept = 7, 
             color = "blue", size=1)+
  ggtitle("Lost Cluster")
gg3 = ggplot(res_melt3, aes(Position, Freq.))+
  geom_bar(stat = "identity", aes(fill = Nucleotide))+
  geom_vline(xintercept = 7, 
             color = "blue", size=1)+
  ggtitle("Stable Cluster")
```

```{r, eval = F}
#all(res_melt1$Nucleotide == res_melt2$Nucleotide)
#all(res_melt2$Nucleotide == res_melt3$Nucleotide)

#all(res_melt1$Position == res_melt2$Position)
#all(res_melt2$Position == res_melt3$Position)


res_tab = cbind(res_melt1, res_melt3[,3], res_melt2[,3])
colnames(res_tab)[3:5] <- c("Gained", "Stable", "Lost")
#pander::pander(res_tab)

#Table 1: the numbers associated with Fig 2.
```



```{r, fig.width=18, fig.height=7}
figure <- ggarrange(gg1, gg3, gg2,
                    ncol = 3, nrow = 1)
figure
```

Fig 2:  Nucleotide distribution in a 14bp window centered at the mid-point of the canonical CTCF motif for binding sequences in cluster Gained, stable and lost. The vertical blue line indicates the center which corresponds to the the 10th nucleotide ('G') in Fig 1.

```{r, fig.width = 6, fig.height =3}

#seqLogo(res_revers) 
seqLogo(res1)
#seqLogo(motif11@pwm)
#mtext("Gained Cluster", 3)
#Fig 2: sequence logo of the CTCF motif identifed in the **Gained** Cluster
```

Fig 3: sequence logo of the CTCF motif identifed in the **Gained** Cluster


```{r, fig.width = 6, fig.height =3}
#seqLogo(motif31@pwm[, 3:13])
seqLogo(res3)
```

Fig 4: sequence logo of the CTCF motif identifed in the **Stable** Cluster

```{r, fig.width = 6, fig.height =3}
seqLogo(res2)
#seqLogo(motif21@pwm[,1:11])
#
```

Fig 5: sequence logo of the CTCF motif identifed in the **Loss** Cluster



# Comparing the identified CTCF-like consensus in the three cluster, in a **21 bp** window

In this section, I will look into the identified CTCF-like censensus in the *Gained*, *Lost*, and *Stable* cluster and compare their binding site orientation, and their nucleotide distribution in a longer window (21bp) centered at the mid-point of the canonical CTCF motif, i.e the 10th nucleotide ('G') in the CTCF logo in Fig. 1.

```{r}
# define the GRange object
# Object out.ranges stores all the genomic Ranges -- a total of 78601 ranges
out.ranges<- readRDS("/scratch/greenwood/kaiqiong.zhao/kaiqiong.zhao/Projects/WitcherKI/Results/out_ranges_June_28.rds")
out_order <- out.ranges[order(out.ranges$PValue),]

# The top 100 down/up sequences

out_down <-  out_order[out_order$direction =="down",]
out_up <-  out_order[out_order$direction =="up",]

gr_up <- out_up[1:1000,]
gr_down <- out_down[1:1000, ]
gr_null <- out_order[(length(out_order)-999):length(out_order),]




```

```{r}

Seqs_up=Biostrings::getSeq(Hsapiens, GRanges(gr_up) + 60 )
Seqs_down=Biostrings::getSeq(Hsapiens, GRanges(gr_down) + 60 )
Seqs_null=Biostrings::getSeq(Hsapiens, GRanges(gr_null) +60 )  

# in the paper "CTCF maintains regulatory homeostasis of cancer pathways", authors extracted the 500 bp DNA sequences centered at the mid-point of the regions defined in the csaw analysis

# Q: do we need to grow the sequence up to 500 bps?

# Yes (because, if we want to extend the sequence plus 20bp in both sides, some might encounter the boundary)

#see1 = GRanges(seqs_up) + 60
```

```{r, eval=F}
# define a function to extract a set of sequences of length bp.use and calculate the Nucleotide probability

seq_set_extract <- function(rr1, motif11, Seqs_up,bp.use = 21){
  ext = (bp.use-1)/2 - 5
  #if(width(rr1$seq_vec) %>% unique()==13){  # up sequences identified by rGADEM
    start = 2; end = 14; center = 8
  #}
  #if(width(rr1$seq_vec) %>% unique() ==18){ # down sequences
  #  start = 1; end = 11; center = 6
  #}
  # if(width(rr1$seq_vec) %>% unique() ==14){ # null sequences
  #  start = 3; end = 13; center = 8
 # }
  seqs_set <- rep(NA, length(rr1$strand_vec))
  
  for ( i in 1:length(rr1$strand_vec)){
    if(rr1$strand_vec[i] == "+"){
      seq_nn <- Seqs_up[motif11@alignList[[i]]@seqID]
  #rr1$seq_vec[i]
oo = gregexpr(DNAString( rr1$seq_vec[i]), seq_nn)

seqs_set[i]<- subseq(seq_nn , start= oo[[1]][1] + (start-1) - ext, end = oo[[1]][1]+(end-1) + ext )
      
     # id_now = rr1$seqID_vec[i]
   #seqs_set[i]<- subseq(Seqs_up[id_now],
    ##   start= motif11@alignList[[i]]@pos+(start-1) - ext ,
     #  motif11@alignList[[i]]@pos+(end-1) + ext  )
    }
    if(rr1$strand_vec[i] == "-"){
      seq_nn <- Seqs_up[motif11@alignList[[i]]@seqID]

oo = gregexpr(DNAString( rr1$seq_vec[i])%>% reverseComplement(), seq_nn)
if(oo[[1]][1] != -1){

seqs_set[i]<-subseq(seq_nn %>% reverseComplement(), start= width(seq_nn)-oo[[1]][1]+1-10 -ext-(width(rr1$seq_vec[i])-end) , width(seq_nn)-oo[[1]][1]+1 + ext -(width(rr1$seq_vec[i])-end) )
  

# if(width(rr1$seq_vec) %>% unique()==18){
   
 #  subseq(seq_nn , start= oo[[1]][1] + (start-1) - ext, end = oo[[1]][1]+(end-1) + ext ) %>% reverseComplement()
#seqs_set[i]<-subseq(seq_nn %>% reverseComplement(), start= width(seq_nn)-oo[[1]][1], width(seq_nn)-oo[[1]][1]+1+11 )
#}

}
    }
    
  }
  seqs_set <- seqs_set[!is.na(seqs_set)]
  res =consensusMatrix(DNAStringSet(seqs_set ), as.prob = TRUE)[1:4,]
res_melt1<-melt(res)
colnames(res_melt1)<- c("Nucleotide", "Position", "Freq.")
return(list(res = res, res_melt = res_melt1))
}

```

```{r, eval=F}

#rr1$seq_vec
res_revers =consensusMatrix(reverseComplement(DNAStringSet(rr1$seq_vec)), as.prob = TRUE)[1:4,]
seqLogo(res_revers)

```


```{r a longer seq_set_extract function}
seq_set_extract <- function(rr1, motif11, Seqs_up,bp.use = 21, reverse = FALSE, rrWidth=13, start = 2, end = 14){
  ext = (bp.use-1)/2 - (rrWidth-1)/2
  #if(width(rr1$seq_vec) %>% unique()==13){  # up sequences identified by rGADEM
  
  #}
  #if(width(rr1$seq_vec) %>% unique() ==18){ # down sequences
  #  start = 1; end = 11; center = 6
  #}
   #if(width(rr1$seq_vec) %>% unique() ==14){ # null sequences
  #  start = 3; end = 13; center = 8
  #}
  seqs_set <- rep(NA, length(rr1$strand_vec))
  
  for ( i in 1:length(rr1$strand_vec)){
    if(rr1$strand_vec[i] == "+"){
      seq_nn <- Seqs_up[motif11@alignList[[i]]@seqID] #the raw input long sequence
  #rr1$seq_vec[i]
oo = gregexpr(DNAString( rr1$seq_vec[i]), seq_nn)

seqs_set[i]<- subseq(seq_nn , start= oo[[1]][1] + (start-1) - ext, end = oo[[1]][1]+(end-1) + ext )
      
     # id_now = rr1$seqID_vec[i]
   #seqs_set[i]<- subseq(Seqs_up[id_now],
    ##   start= motif11@alignList[[i]]@pos+(start-1) - ext ,
     #  motif11@alignList[[i]]@pos+(end-1) + ext  )
    }
    if(rr1$strand_vec[i] == "-"){
      seq_nn <- Seqs_up[motif11@alignList[[i]]@seqID] #the raw input long sequence -- this is in the reverse strand

#oo = gregexpr(DNAString( rr1$seq_vec[i])%>% reverseComplement(), seq_nn)
#if(oo[[1]][1] != -1){

#seqs_set[i]<-subseq(seq_nn %>% reverseComplement(), start= width(seq_nn)-oo[[1]][1]+1-10 -ext-(width(rr1$seq_vec[i])-end) , #width(seq_nn)-oo[[1]][1]+1 + ext -(width(rr1$seq_vec[i])-end) )


# Modify the reverse strand extraction-- July 1 reverse the original input strand instead --> advantage: less tuning for start and end for the - strand

oo = gregexpr(DNAString(rr1$seq_vec[i]), seq_nn %>% reverseComplement())
if(oo[[1]][1] != -1){

seqs_set[i]<- subseq(seq_nn %>% reverseComplement() , start= oo[[1]][1] + (start-1) - ext, end = oo[[1]][1]+(end-1) + ext )



# if(width(rr1$seq_vec) %>% unique()==18){
   
 #  subseq(seq_nn , start= oo[[1]][1] + (start-1) - ext, end = oo[[1]][1]+(end-1) + ext ) %>% reverseComplement()
#seqs_set[i]<-subseq(seq_nn %>% reverseComplement(), start= width(seq_nn)-oo[[1]][1], width(seq_nn)-oo[[1]][1]+1+11 )
#}

}
    }
    
  }
  seqs_set <- seqs_set[!is.na(seqs_set)]
  
  if(reverse){
    res =consensusMatrix( reverseComplement(DNAStringSet(seqs_set )), as.prob = TRUE)[1:4,]
  }else{
res =consensusMatrix(DNAStringSet(seqs_set ), as.prob = TRUE)[1:4,]
  }
  res_melt1<-melt(res)
colnames(res_melt1)<- c("Nucleotide", "Position", "Freq.")
return(list(res = res, res_melt = res_melt1, seqs_set=seqs_set)) 
# April 5, 2021, extracting the raw sequence matrix as well
}
```


```{r}
# in the cluster "Down", there are 2 sequences without the strand information
res_ext1 <- seq_set_extract(rr1, motif11, Seqs_up, bp.use = 61, reverse = TRUE, rrWidth = 13, start = 1, end = 13)
#-------
# for the up cluster -- the orignal rGADEM identified sequences of length 1-14, center is on position 8
#----------------

# start: start in the original subsequences rr1$seq_vec[i]
# end: end in the original subsequences rr1$seq_vec[i]
# make everything symetric; facilitate any follow-up manipulation for the reverse strand

# For the reverse = TRUE sequences, carefully select the start and end

res_ext2 <- seq_set_extract(rr2, motif21, Seqs_down, bp.use = 61, reverse = TRUE, rrWidth = 13, start = 3, end = 15)

# Down cluster -- 16 bp, center is on position 8

res_ext3<- seq_set_extract(rr3, motif31, Seqs_null, bp.use = 61, reverse = FALSE,
                           rrWidth = 13, start = 2, end = 14)
```




```{r}
res1 = res_ext1$res_melt
res2 = res_ext2$res_melt
res3 = res_ext3$res_melt



res1_21<- res1[res1$Position  %in% 21:41,] # 21bp
res2_21<- res2[res2$Position  %in% 21:41,] # 21bp
res3_21<- res3[res3$Position  %in% 21:41,] # 21bp

res1_21$Position = res1_21$Position-20
res2_21$Position = res2_21$Position-20
res3_21$Position = res3_21$Position-20

res_tab = cbind(res1_21, res3_21[,3], res2_21[,3])
colnames(res_tab)[3:5] <- c("Gained", "Stable", "Lost")
rownames(res_tab) = NULL
#pander::pander(res_tab)

```


```{r, fig.width=18, fig.height=6}
res_span <- melt(res_tab, id.vars = c("Nucleotide", "Position"))
colnames(res_span)[3:4] <- c("Type", "Freq.")

ggplot(res_span, aes(Position, Freq.))+
  geom_bar(stat = "identity", aes(fill = Nucleotide))+
  geom_vline(xintercept = 11, 
             color = "blue", size=1)+
  facet_wrap(~Type, nrow = 1)
```


Fig 6: Necleotide distribution in a 21bp window centered at the mid-point of the canonical CTCF motif for binding sequences in cluster Gained, stable and lost. The vertical blue line indicates the center which corresponds to the 10th nucleotide ('G') in Fig 1.


```{r, fig.width = 8, fig.height =3}
seqLogo( res_ext1$res[, 21:41])

```


Fig 7: sequence logo of the CTCF motif identifed in the **Gained** Cluster

```{r, fig.width = 8, fig.height =3}

seqLogo( res_ext3$res[, 21:41])


```

Fig 8: sequence logo of the CTCF motif identifed in the **Stable** Cluster

```{r, fig.width = 8, fig.height =3}

seqLogo( res_ext2$res[, 21:41])

```

Fig 9: sequence logo of the CTCF motif identifed in the **Lost** Cluster


# Comparing the identified CTCF-like consensus in the three cluster, in a **41 bp** window


```{r}

res1_21<- res1[res1$Position  %in% 11:51,] # 61bp
res2_21<- res2[res2$Position  %in% 11:51,] # 61bp
res3_21<- res3[res3$Position  %in% 11:51,] # 61bp

res1_21$Position = res1_21$Position-10
res2_21$Position = res2_21$Position-10
res3_21$Position = res3_21$Position-10

res_tab = cbind(res1_21, res3_21[,3], res2_21[,3])
colnames(res_tab)[3:5] <- c("Gained", "Stable", "Lost")
rownames(res_tab) = NULL
#pander::pander(res_tab)

```


```{r, fig.width=18, fig.height=6}
res_span <- melt(res_tab, id.vars = c("Nucleotide", "Position"))
colnames(res_span)[3:4] <- c("Type", "Freq.")

ggplot(res_span, aes(Position, Freq.))+
  geom_bar(stat = "identity", aes(fill = Nucleotide))+
  geom_vline(xintercept = 21, 
             color = "blue", size=1)+
  facet_wrap(~Type, nrow = 1)
```


Fig 10: Necleotide distribution in a 61bp window centered at the mid-point of the canonical CTCF motif for binding sequences in cluster Gained, stable and lost. The vertical blue line indicates the center which corresponds to the 10th nucleotide ('G') in Fig 1.


```{r, fig.width = 18, fig.height =3}
seqLogo( res_ext1$res[, 11:51])

```


Fig 11: sequence logo of the CTCF motif identifed in the **Gained** Cluster

```{r, fig.width = 18, fig.height =3}

seqLogo( res_ext3$res[, 11:51])


```

Fig 12: sequence logo of the CTCF motif identifed in the **Stable** Cluster

```{r, fig.width = 18, fig.height =3}

seqLogo( res_ext2$res[, 11:51])

```

Fig 13: sequence logo of the CTCF motif identifed in the **Lost** Cluster

# Comparing the identified CTCF-like consensus in the three cluster, in a **61 bp** window


```{r}

res1_21<- res1
res2_21<- res2
res3_21<- res3



res_tab = cbind(res1_21, res3_21[,3], res2_21[,3])
colnames(res_tab)[3:5] <- c("Gained", "Stable", "Lost")
rownames(res_tab) = NULL
#pander::pander(res_tab)

```


```{r, fig.width=18, fig.height=6}
res_span <- melt(res_tab, id.vars = c("Nucleotide", "Position"))
colnames(res_span)[3:4] <- c("Type", "Freq.")

ggplot(res_span, aes(Position, Freq.))+
  geom_bar(stat = "identity", aes(fill = Nucleotide))+
  geom_vline(xintercept = 31, 
             color = "blue", size=1)+
  facet_wrap(~Type, nrow = 1)
```


Fig 14: Necleotide distribution in a 61bp window centered at the mid-point of the canonical CTCF motif for binding sequences in cluster Gained, stable and lost. The vertical blue line indicates the center which corresponds to the 10th nucleotide ('G') in Fig 1.


```{r, fig.width = 18, fig.height =3}
seqLogo( res_ext1$res)

```


Fig 15: sequence logo of the CTCF motif identifed in the **Gained** Cluster

```{r, fig.width = 18, fig.height =3}

seqLogo( res_ext3$res)


```

Fig 16: sequence logo of the CTCF motif identifed in the **Stable** Cluster

```{r, fig.width = 18, fig.height =3}

seqLogo( res_ext2$res)

```

Fig 17: sequence logo of the CTCF motif identifed in the **Lost** Cluster

# Frequency differences


## **Lost** v.s. **Stable**

```{r}

res_tab$`lost-stable` <- res_tab$Lost - res_tab$Stable
res_tab$`Gained-stable` <- res_tab$Gained - res_tab$Stable
setwd("~/scratch/kaiqiong.zhao/Projects/WitcherKI/3-Extend-motifs")
write.csv(res_tab, "KI_freqTab_61bp.csv")

ggplot(res_tab, aes(Position, `lost-stable`))+
  geom_point(stat = "identity" , aes(colour = Nucleotide, shape = Nucleotide))+
  geom_vline(xintercept = 31, 
             color = "blue", size=0.5, linetype = 3)+
    geom_hline(yintercept = 0.1, linetype = 2, size=0.5, color = "gray") +
     geom_hline(yintercept = -0.1, linetype = 2, size=0.5, color = "gray") +
  geom_vline(xintercept = 25, linetype = 2, size=0.5) +
    geom_vline(xintercept = 48, linetype = 2, size=0.5) +
  ggtitle("Lost cluster v.s Stable Cluster")+
  ylab("Freq. differences")
```

Fig 18: Difference of nucleotide differences between the **Lost** and **Stable** clusters. Black dashed lines (at position 25 and 48) are the boundaries of the region with frequency differences greater than 10% between the Lost cluster and the stable cluster

Table: Nucleotide frequences at positions with Freq. differences greater than 10% between the **Lost** and **Stable** clusters.

```{r}
see = order(abs(res_tab$`lost-stable`), decreasing = T)
res_tab_ord = res_tab[see,]
ss = res_tab_ord
rownames(ss) =NULL
ss = ss[abs(ss$`lost-stable`)>0.1,]
pander::pander(ss , split.table=Inf)
```




## **Gained** v.s. **Stable**

```{r}
ggplot(res_tab, aes(Position, `Gained-stable`))+
  geom_point(stat = "identity" , aes(colour = Nucleotide, shape = Nucleotide))+
 geom_vline(xintercept = 31, 
             color = "blue", size=0.5, linetype = 3)+
    geom_hline(yintercept = 0.1, linetype = 2, size=0.5, color = "gray") +
     geom_hline(yintercept = -0.1, linetype = 2, size=0.5, color = "gray") +
  geom_vline(xintercept = 33, linetype = 2, size=0.5) +
    geom_vline(xintercept = 47, linetype = 2, size=0.5) +
  ggtitle("Gained cluster v.s Stable Cluster")+
  ylab("Freq. differences")+
  coord_cartesian(ylim = c(-0.2, 0.45))
```

Fig 19: Difference of nucleotide differences between the **Lost** and **Stable** clusters. Black dashed lines (at position 33 and 47) are the boundaries of the region with frequency differences greater than 10% between the Gained cluster and the Stable cluster.

Table: Nucleotide frequences at positions with Freq. differences greater than 10% between the Gained cluster and the Stable cluster.


```{r}
see = order(abs(res_tab$`Gained-stable`), decreasing = T)
res_tab_ord = res_tab[see,]
ss = res_tab_ord
rownames(ss) =NULL
ss = ss[abs(ss$`Gained-stable`)>0.1,]
pander::pander(ss , split.table=Inf)

```

## Nucleotide frequencies in a **24 bp** window (this region has Freq. differences greater than 10%)




```{r}

res1_21<- res1[res1$Position  %in% 25:48,] # 21bp
res2_21<- res2[res2$Position  %in% 25:48,] # 21bp
res3_21<- res3[res3$Position  %in% 25:48,] # 21bp

res1_21$Position = res1_21$Position-24
res2_21$Position = res2_21$Position-24
res3_21$Position = res3_21$Position-24



res_tab = cbind(res1_21, res3_21[,3], res2_21[,3])
colnames(res_tab)[3:5] <- c("Gained", "Stable", "Lost")
rownames(res_tab) = NULL
#pander::pander(res_tab)

```


```{r, fig.width=12, fig.height=4}
res_span <- melt(res_tab, id.vars = c("Nucleotide", "Position"))
colnames(res_span)[3:4] <- c("Type", "Freq.")

ggplot(res_span, aes(Position, Freq.))+
  geom_bar(stat = "identity", aes(fill = Nucleotide))+
  geom_vline(xintercept = 31-24, 
             color = "blue", size=1)+
  facet_wrap(~Type, nrow = 1)
```


Fig 20: Necleotide distribution in a window for binding sequences in cluster Gained, stable and lost. The vertical blue line indicates the center which corresponds to the 10th nucleotide ('G') in Fig 1.


```{r, fig.width = 18, fig.height =5}
seqLogo( res_ext1$res[,25:48])

```


Fig 21: sequence logo of the CTCF motif identifed in the **Gained** Cluster

```{r, fig.width = 18, fig.height =5}

seqLogo( res_ext3$res[,25:48])


```

Fig 22: sequence logo of the CTCF motif identifed in the **Stable** Cluster

```{r, fig.width = 18, fig.height =5}

seqLogo( res_ext2$res[,25:48])

```

Fig 23: sequence logo of the CTCF motif identifed in the **Lost** Cluster



```{r, eval = F}
library(compositions)

?acomp

sss = acomp(t(res_ext1$res)[26:36,])
plot(sss, pch = 19)
```



```{r, eval=TRUE}
knitr::knit_exit()
```
